LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m

```diff
@@ -0,0 +1,233 @@
+    BOOL isDirectory = NO;
+    if (![[NSFileManager defaultManager] fileExistsAtPath:directoryPath isDirectory:&isDirectory]) {
+        [[NSFileManager defaultManager] createDirectoryAtPath:directoryPath withIntermediateDirectories:YES attributes:nil error:nil];
+    }
+}
+
+/**
+ Returns the path to the directory to save snapshots of the current failing test. Path includes class and method name
+ e.g. {FULL_PATH}/SamepleTableViewCellLayoutTests/testSampleTableViewCell
+ */
+- (NSString *)directoryPathForCurrentInvocation:(NSInvocation *)invocation {
+    NSString *documentsDirectory = [self commonRootPath];
+    
+    NSString *directoryPath = [documentsDirectory stringByAppendingString:[NSString stringWithFormat:@"/%@", [self methodNameForInvocation:invocation]]];
+    return directoryPath;
+}
+
+- (NSString *)methodNameForInvocation:(NSInvocation *)invocation {
+    return NSStringFromSelector((SEL)[invocation selector]);
+}
+
+- (NSString *)commonRootPath {
+    return [self commonRootPathForInvocationClass:self.invocationClass];
+}
+
+- (NSString *)commonRootPathForInvocationClass:(Class)classForRoot {
+    NSString *currentDirectory = [[NSBundle bundleForClass:classForRoot] bundlePath];
+    NSString *className = NSStringFromClass(classForRoot);
+    //Check incase the class name includes a ".", if so we the actual class name will be everything after the "."
+    if ([className containsString:@"."]) {
+        className = [className componentsSeparatedByString:@"."].lastObject;
+    }
+    
+    return [currentDirectory stringByAppendingPathComponent:[NSString stringWithFormat:@"LayoutTestImages/%@", className]];
+}
+
+/**
+ Returns the full path to the image with the given file name and the iamge size and width appended to it.
+ e.g. {DIRECTORY_PATH}/SamepleTableViewCellLayoutTests/testSampleTableViewCell/{FILE_NAME}_{IMAGE_WIDTH}_{IMAGE_HEIGHT}.png
+ */
+- (NSString *)pathForImageWithWidth:(CGFloat)width height:(CGFloat)height data:(NSDictionary *)data invocation:(NSInvocation *)invocation {
+    NSString *directoryPath = [self directoryPathForCurrentInvocation:invocation];
+    NSString *imageName = [self nameForImageWithWidth:width height:height data:data];
+    return [directoryPath stringByAppendingPathComponent:imageName];
+}
+
+- (NSString *)nameForImageWithWidth:(CGFloat)width height:(CGFloat)height data:(NSDictionary *)data {
+    NSString *imageName = [NSString stringWithFormat:@"Width-%.2f_Height-%.2f_Data-%lu", width, height, (unsigned long)data.description.hash];
+    return [imageName stringByAppendingString:@".png"];
+}
+
+- (UIImage *)renderLayer:(CALayer *)layer {
+    UIImage *snapshot = nil;
+        CGRect bounds = layer.bounds;
+    if (CGRectGetWidth(bounds) > 0 && CGRectGetHeight(bounds) > 0) {
+        UIGraphicsBeginImageContextWithOptions(bounds.size, NO, 0);
+        CGContextRef context = UIGraphicsGetCurrentContext();
+        
+        CGContextSaveGState(context);
```

Comment:
I think we want to wrap this in an if (context) statement? I think this only happens if the bounds has no area, but maybe it can happen for otehr reasons too? Not sure if this will crash.